client/fingerprint/consul.go

```diff
@@ -16,108 +18,214 @@ const (
 
 		f.client, err = consul.NewClient(consulConfig)
 		if err != nil {
-			return fmt.Errorf("Failed to initialize consul client: %s", err)
+			return fmt.Errorf("failed to initialize Consul client: %s", err)
+		}
+
+		f.extractors = map[string]consulExtractor{
+			"consul.server":        f.server,
+			"consul.version":       f.version,
+			"consul.sku":           f.sku,
+			"consul.revision":      f.revision,
+			"unique.consul.name":   f.name,
+			"consul.datacenter":    f.dc,
+			"consul.segment":       f.segment,
+			"consul.connect":       f.connect,
+			"consul.grpc":          f.grpc,
+			"consul.ft.namespaces": f.namespaces,
 		}
 	}
 
+	return nil
+}
+
+func (f *ConsulFingerprint) query(resp *FingerprintResponse) consulInfo {
 	// We'll try to detect consul by making a query to to the agent's self API.
 	// If we can't hit this URL consul is probably not running on this machine.
 	info, err := f.client.Agent().Self()
 	if err != nil {
 		f.clearConsulAttributes(resp)
 
-		// Print a message indicating that the Consul Agent is not available
-		// anymore
+		// indicate consul no longer available
 		if f.lastState == consulAvailable {
 			f.logger.Info("consul agent is unavailable")
 		}
 		f.lastState = consulUnavailable
 		return nil
 	}
+	return info
+}
 
-	if s, ok := info["Config"]["Server"].(bool); ok {
-		resp.AddAttribute("consul.server", strconv.FormatBool(s))
-	} else {
-		f.logger.Warn("unable to fingerprint consul.server")
-	}
-	if v, ok := info["Config"]["Version"].(string); ok {
-		resp.AddAttribute("consul.version", v)
-	} else {
-		f.logger.Warn("unable to fingerprint consul.version")
-	}
-	if r, ok := info["Config"]["Revision"].(string); ok {
-		resp.AddAttribute("consul.revision", r)
-	} else {
-		f.logger.Warn("unable to fingerprint consul.revision")
-	}
-	if n, ok := info["Config"]["NodeName"].(string); ok {
-		resp.AddAttribute("unique.consul.name", n)
-	} else {
-		f.logger.Warn("unable to fingerprint unique.consul.name")
-	}
-	if d, ok := info["Config"]["Datacenter"].(string); ok {
-		resp.AddAttribute("consul.datacenter", d)
-	} else {
-		f.logger.Warn("unable to fingerprint consul.datacenter")
-	}
-	if g, ok := info["Member"]["Tags"].(map[string]interface{}); ok {
-		if s, ok := g["segment"].(string); ok {
-			resp.AddAttribute("consul.segment", s)
-		}
-	} else {
-		f.logger.Warn("unable to fingerprint consul.segment")
-	}
+func (f *ConsulFingerprint) link(resp *FingerprintResponse) {
 	if dc, ok := resp.Attributes["consul.datacenter"]; ok {
 		if name, ok2 := resp.Attributes["unique.consul.name"]; ok2 {
 			resp.AddLink("consul", fmt.Sprintf("%s.%s", dc, name))
 		}
 	} else {
 		f.logger.Warn("malformed Consul response prevented linking")
 	}
+}
 
-	// If the Consul Agent was previously unavailable print a message to
-	// indicate the Agent is available now
-	if f.lastState == consulUnavailable {
-		f.logger.Info("consul agent is available")
+func (f *ConsulFingerprint) server(info consulInfo) (string, bool) {
```

Comment:
I love how legible this entire `ConsulFingerprint` becomes with these extractor functions. 

This refactoring does make me realize that if Consul were to push a backwards incompatible JSON we could potentially panic Nomad when we do the type cast. Probably something for future robustness work to worry about rather than this PR though.

Reply:
I think we're actually okay (for panics) on type changes - each of these type casts is in the context of a type check which will simply become false. The attributes will end up being _removed_ which isn't great, of course, and I'll probably have a few choice words with Consul team :sweat_smile: